### PR TITLE
Add a generic "interruptable".

### LINKED ISSUE #883: Places should respect interrupt during sync
This is required for iOS bookmark integration, because they start a sync on backgrounding, but need to be able to interrupt it (when the background deadline runs out) so that we can close file handles and such to avoid crashes.

### PR DESCRIPTION
**Pull request by mhammond**
This attempts to expose the ability to interrupt a long running process
in a way that can be used across all of our crates, to fix #883.

It introduces a new crate, support/integration, with this support. The name
'integration' was chosen as this could form the basis for other things shared
across our crates to enable better integration between them. However, this
might be putting the cart before the horse, and it might be better named as
simply 'interrupt'.

It defines 2 traits and a single error. These traits are used whenever the
interrupt concept needs to cross crate boundaries.

Note that this *does not* deeply integrate support for interrupting into
the sync crate for 2 reasons:

* First, to see if the concept itself has legs.

* Thom mentioned it might be possible to use a tls crate to avoid passing
  interrupt stuff everywhere - however, I can't see how to make this work.
  I looked into `scoped-tls` and a few other tls crates, but they all seem
  to make the value available only in a single "scope", so I can see how
  they can help us here. Thom, please let me know what you were thinking.

It adds the `ctrlc` crate, and makes a couple of things `pub` so that the
`places-utils` example can be interruptable in a "nice" way. Both of these
could be considered optional, but they make sense to me.

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - `cargo test --all` produces no test failures
  - `cargo clippy --all --all-targets --all-features` runs without emitting any warnings
  - `cargo fmt` does not produce any changes to the code
  - `./gradlew ktlint detekt` runs without emitting any warnings
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/master/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

File: components/support/integration/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+pub mod error;
+mod interrupt;
+
```
Comment (thomcc) [inline components/support/integration/src/lib.rs]: nit: reexport the error here too.

File: components/logins/Cargo.toml [hunk 2]
```diff
@@ -21,6 +21,7 @@ url = "1.7.1"
 failure = "0.1.3"
 sql-support = { path = "../support/sql" }
 ffi-support = { path = "../support/ffi" }
+integration = { path = "../support/integration" }
```
Comment (thomcc) [inline components/logins/Cargo.toml]: This should probably just be `support/interrupt` or something.

File: components/logins/src/engine.rs [hunk 3]
```diff
@@ -4,10 +4,20 @@
 use crate::db::LoginDb;
 use crate::error::*;
 use crate::login::Login;
+use integration::Interruptee;
 use std::cell::Cell;
 use std::path::Path;
 use sync15::{sync_multiple, telemetry, ClientInfo, KeyBundle, Sync15StorageClientInit};
 
+// For now, logins doesn't support being interrupted.
+struct NeverInterrupts;
```
Comment (thomcc) [inline components/logins/src/engine.rs]: Maybe move this into `support/interrupt` too.

File: components/support/integration/src/interrupt.rs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+// Helps manage "interruptable" things across our various crates.
+
+use crate::error::*;
+
+/// Something that is interruptable. In practice, this will almost certainly
+/// be Sync + Send, as it will typically be created on one thread, but
+/// `.interrupt()` will be called from a different thread. However,
+/// Sync + Send semantics aren't mandated here.
+pub trait Interruptable {
```
Comment (thomcc) [inline components/support/integration/src/interrupt.rs]: This is more general than I think we need (and this generality was why you had trouble with the `'static` bound). That said, for now this is fine, and we can consider simplifying later..

File: components/logins/src/engine.rs [hunk 5]
```diff
@@ -4,10 +4,20 @@
...
+
+impl Interruptee for NeverInterrupts {
+    fn was_interrupted(&self) -> bool {
```
Comment (thomcc) [inline components/logins/src/engine.rs]: nit: `#[inline]` probably.

File: components/places/examples/places-utils.rs [hunk 6]
```diff
@@ -165,6 +166,16 @@ fn sync(
     reset: bool,
 ) -> Result<()> {
     let conn = api.open_sync_connection()?;
+
+    // interrupts are per-connection, so we need to set that up here.
+    let interrupt_handle = conn.new_interrupt_handle();
+
+    ctrlc::set_handler(move || {
```
Comment (thomcc) [inline components/places/examples/places-utils.rs]: Is this just for testing?

**Comment by thomcc** [comment]
Logins blocking shutdown and causing crashes iOS seems like it's a big problem in practice, so this is needed.

**Comment by mhammond** [comment]
See the semi-detailed commit message for 0eab6a1

File: components/places/examples/places-utils.rs (hunk 6 above)
Comment (mhammond) [reply components/places/examples/places-utils.rs]: Yep, exactly

File: components/logins/src/db.rs [hunk 7]
```diff
@@ -25,6 +28,10 @@ use sync15::{
 
 pub struct LoginDb {
     pub db: Connection,
+    // The logins DB has exactly 1 interrupt scope, but we keep a copy of the
+    // Arc with the counter so we can create an interrupt handle.
+    interrupt_counter: Arc<AtomicUsize>,
+    pub interrupt_scope: SqlInterruptScope,
```
Comment (thomcc) [inline components/logins/src/db.rs]: Yeah I don't think the scope should live here.

Instead we should add a new `struct HistoryStore<'a> { &'a LoginsDb, SqlInterruptScope }`

File: components/support/interrupt/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,46 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
```
Comment (thomcc) [inline components/support/interrupt/src/lib.rs]: Add these to follow other crates (unfortunately there's no way to configure this for the whole workspace
```
#![allow(unknown_lints)]
#![warn(rust_2018_idioms)]
```
(Note: The second was added today, if you get merge conflicts for the lint changes, you can resolve in favor of yours, and run `cargo fix --edition-idioms` to resolve for the most part)

File: components/support/sql/src/interrupt.rs [hunk 9]
```diff
@@ -14,52 +15,74 @@ use std::sync::{
...
+}
+
+impl Interruptable for SqlInterruptHandle {
+    fn interrupt(&self) {
         self.interrupt_counter.fetch_add(1, Ordering::SeqCst);
         self.db_handle.interrupt();
     }
 }
 
+// Ideally we'd just use implement_into_ffi_by_pointer! from ffi_support, but
```
Comment (thomcc) [inline components/support/sql/src/interrupt.rs]: `implement_into_ffi_by_pointer!` should work here. It just expands into more or less what you wrote. The issue is if we wanted to run it in a crate other than the one that defined this.

**Pull request merged**